node/src/components/linear_chain.rs

```diff
@@ -534,19 +532,34 @@ where
                             .announce_finality_signature(fs.clone())
                             .ignore(),
                     );
-                    block.append_proof(fs.public_key, fs.signature);
+                    signature.append_proof(fs.public_key, fs.signature);
                     // Cache the results in case we receive the same finality signature before we
                     // manage to store it in the database.
-                    self.block_cache.insert(*block.clone());
-                    effects.extend(effect_builder.put_block_to_storage(block).ignore());
+                    self.signature_cache.insert(*signature.clone());
+                    info!(hash=%signature.block_hash, "storing finality signatures");
+                    effects.extend(
+                        effect_builder
+                            .put_signatures_to_storage(*signature)
+                            .ignore(),
+                    );
                     effects
                 }
             }
-            Event::IsBonded(None, _, true) => {
+            Event::IsBonded(None, fs, true) => {
                 // Unknown block but validator is bonded.
                 // We should finalize the same block eventually. Either in this or in the
                 // next era.
-                Effects::new()
+                // Store the signatures.
+                let mut effects = Effects::new();
+                let mut signatures = BlockSignatures::new(fs.block_hash, fs.era_id);
+                signatures.append_proof(fs.public_key, fs.signature);
+                effects.extend(
+                    effect_builder
+                        .put_signatures_to_storage(signatures.clone())
+                        .ignore(),
+                );
+                self.signature_cache.insert(signatures);
```

Comment:
This is changing the logic and is not correct. `signature_cache` should only contain signatures that we have validated and stored. You should remove the whole bit that you added to this match arm.

Reply:
Removed incorrect logic